### PR TITLE
Chachapoly impl and tests

### PR DESCRIPTION
**Pull request by davisgu**

### CODE REVIEW

**Comment by davisgu** [comment]
@matthewstevenson88

File: security/s2a/internal/crypter/aesgcm_test.go [hunk 1]
```diff
@@ -21,8 +21,9 @@ package crypter
 import (
 	"bytes"
 	"fmt"
-	"google.golang.org/grpc/security/s2a/internal/crypter/testutil"
 	"testing"
+
+	"google.golang.org/grpc/security/s2a/internal/crypter/testutil"
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/aesgcm_test.go:26]: Why did these get rearranged?

File: security/s2a/internal/crypter/chachapoly.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+import (
+	"golang.org/x/crypto/chacha20poly1305"
+
+	"crypto/cipher"
+	"fmt"
+)
+
+// Supported key sizes in bytes.
+const (
+	chachaKeySize256 = 32
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/chachapoly.go:30]: Please call this `chacha20Poly1305KeySize`.

File: security/s2a/internal/crypter/chachapoly.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+)
+
+// chachapoly is the struct that holds an CHACHA-POLY cipher for the S2A AEAD crypter.
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/chachapoly.go:33]: s/an/a

File: security/s2a/internal/crypter/aesgcm_test.go (hunk 1 above)
Comment (davisgu) [reply security/s2a/internal/crypter/aesgcm_test.go:26]: interestingly, it seems that emacs automatically rearranges them when I save. I'll try to find a way to prevent that.

File: security/s2a/internal/crypter/chachapoly.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+type chachapoly struct {
+	aead cipher.AEAD
+	// keySize stores the size of the key which was initially used. This
+	// is necessary to restrict key updates to the same key length as the
+	// initial key.
+	keySize int
+}
+
+// newCHACHAPOLY creates an CHACHA-POLY crypter instance. Note that the key must be
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/chachapoly.go:42]: s/an/a

File: security/s2a/internal/crypter/chachapoly.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+// 256 bits.
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/chachapoly.go:43]: s/256 bits/32 bytes

Comment (matthewstevenson88) [reply security/s2a/internal/crypter/chachapoly.go:43]: Since all of our data is in bytes (and not bits), we should make all references in bytes.

File: security/s2a/internal/crypter/chachapoly_test.go [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+
+import (
+	"fmt"
+	"testing"
+
+	"google.golang.org/grpc/security/s2a/internal/crypter/testutil"
+)
+
+// getGCMCryptoPair outputs a sender/receiver pair on CHACHA-POLY.
+func getPOLYCryptoPair(key []byte, t *testing.T) (s2aAeadCrypter, s2aAeadCrypter) {
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/chachapoly_test.go:29]: s/POLYCrypto/ChachaPolyCrypter

File: security/s2a/internal/crypter/chachapoly_test.go [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+package crypter
+
+import (
+	"fmt"
+	"testing"
+
+	"google.golang.org/grpc/security/s2a/internal/crypter/testutil"
+)
+
+// getGCMCryptoPair outputs a sender/receiver pair on CHACHA-POLY.
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/chachapoly_test.go:28]: Please change the `getGCMCryptoPair` part of the comment.

File: security/s2a/internal/crypter/chachapoly.go [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+func newCHACHAPOLY(key []byte) (s2aAeadCrypter, error) {
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/chachapoly.go:44]: Please make this camel-case, e.g. `newChachaPoly`.

Comment (davisgu) [reply security/s2a/internal/crypter/chachapoly.go:44]: --should I make all the test functions the same?

File: security/s2a/internal/crypter/chachapoly.go [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+	// If dst is equal to ciphertext[:0], ciphertext storage is reused.
+	plaintext, err := s.aead.Open(dst, nonce, ciphertext, aad)
+	if err != nil {
+		return nil, fmt.Errorf("message auth failed: %v", err)
+	}
+	return plaintext, nil
+}
+
+func (s *chachapoly) tagSize() int {
+	return overhead
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/chachapoly.go:93]: What is `overhead`? Where is it set?

File: security/s2a/internal/crypter/chachapoly.go [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+}
+
+func (s *chachapoly) updateKey(key []byte) error {
+	if s.keySize != len(key) {
+		return fmt.Errorf("supplied key must have same size as initial key: %d bits", s.keySize*8)
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/chachapoly.go:98]: Change from bits to bytes.

File: security/s2a/internal/crypter/chachapoly_test.go [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+	return sender, receiver
+}
+
+func wycheProofTestVectorFilterCCP(testGroup testutil.TestGroup) bool {
+	return testGroup.IVSize != 96 ||
+		(testGroup.KeySize != 256) ||
+		testGroup.TagSize != 128
+}
+
+func testPOLYEncryptionDecryption(sender s2aAeadCrypter, receiver s2aAeadCrypter, test *testutil.CryptoTestVector, t *testing.T) {
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/chachapoly_test.go:47]: s/POLY/ChachaPoly

File: security/s2a/internal/crypter/chachapoly_test.go [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+		t.Fatalf("newCHACHAPOLY(ClientSide, key) = %v", err)
+	}
+	receiver, err := newCHACHAPOLY(key)
+	if err != nil {
+		t.Fatalf("newCHACHAPOLY(ServerSide, key) = %v", err)
+	}
+	return sender, receiver
+}
+
+func wycheProofTestVectorFilterCCP(testGroup testutil.TestGroup) bool {
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/chachapoly_test.go:41]: s/CCP/ChachaPoly

File: security/s2a/internal/crypter/chachapoly_test.go [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+
+	// Decrypt.
+	got, err = receiver.decrypt(nil, ciphertext, test.Nonce, test.Aad)
+	if isFailure(test.Result, err, got, test.Plaintext) {
+		t.Errorf("key=%v\nDecrypt(\n dst = nil\n ciphertext = %v\n nonce = %v\n aad = %v\n) = (\n %v\n %v\n), want %v",
+			test.Key, ciphertext, test.Nonce, test.Aad, got, err, test.Plaintext)
+	}
+}
+
+func testPOLYEncryptRoundtrip(sender s2aAeadCrypter, receiver s2aAeadCrypter, t *testing.T) {
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/chachapoly_test.go:70]: s/POLY/ChachaPoly

File: security/s2a/internal/crypter/chachapoly_test.go [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+	if err != nil {
+		t.Fatalf("Decrypt(%v, %v, %v, nil) failed, err = %v", ciphertext[:0], ciphertext, nonce, err)
+	}
+	if string(decryptedPlaintext) != plaintext {
+		t.Fatalf("Decrypt(%v, %v, %v, nil) = %v, want %v", ciphertext[:0], ciphertext, nonce, decryptedPlaintext, plaintext)
+	}
+}
+
+// Test encrypt and decrypt using an invalid key size.
+func TestCHACHAPOLYInvalidKeySize(t *testing.T) {
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/chachapoly_test.go:94]: s/CHACHAPOLY/ChachaPoly

File: security/s2a/internal/crypter/chachapoly_test.go [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+	// Use 17 bytes, which is invalid.
+	key := make([]byte, 17)
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/chachapoly_test.go:96]: Instead of 17, can you use the key size constant + 1?

File: security/s2a/internal/crypter/chachapoly_test.go [hunk 16]
```diff
@@ -0,0 +1,214 @@
...
+	if _, err := newCHACHAPOLY(key); err == nil {
+		t.Error("expected an error when using invalid key size")
+	}
+}
+
+// Test update key for CHACHA-POLY  using a key with different size from the initial
+// key.
+func TestCHACHAPOLYKeySizeUpdate(t *testing.T) {
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/chachapoly_test.go:104]: s/CHACHAPOLY/ChachaPoly

File: security/s2a/internal/crypter/chachapoly_test.go [hunk 17]
```diff
@@ -0,0 +1,214 @@
...
+			newKey := make([]byte, tc.updateKeySize)
+			if err = crypter.updateKey(newKey); err == nil {
+				t.Fatal("UpdateKey should fail with invalid key size error")
+			}
+		})
+	}
+}
+
+// Test Encrypt/Decrypt using an invalid nonce size.
+func TestCHACHAPOLYEncryptDecryptInvalidNonce(t *testing.T) {
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/chachapoly_test.go:128]: Same naming style here.

File: security/s2a/internal/crypter/chachapoly_test.go [hunk 18]
```diff
@@ -0,0 +1,214 @@
...
+	if _, err = crypter.encrypt(nil, nil, nonce, nil); err == nil {
+		t.Errorf("Encrypt should fail due to invalid nonce size")
+	}
+	if _, err = crypter.decrypt(nil, nil, nonce, nil); err == nil {
+		t.Fatalf("Decrypt should fail due to invalid nonce size")
+	}
+}
+
+// Test encrypt and decrypt on roundtrip messages for CHACHA-POLY.
+func TestCHACHAPOLYEncryptRoundtrip(t *testing.T) {
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/chachapoly_test.go:145]: Same naming style here.

File: security/s2a/internal/crypter/chachapoly_test.go [hunk 19]
```diff
@@ -0,0 +1,214 @@
...
+	for _, keySize := range []int{chachaKeySize256} {
+		key := make([]byte, keySize)
+		sender, receiver := getPOLYCryptoPair(key, t)
+		testPOLYEncryptRoundtrip(sender, receiver, t)
+	}
+}
+
+// Test encrypt and decrypt on roundtrip messages for CHACHA-POLY using an updated
+// key.
+func TestCHACHAPOLYUpdatedKey(t *testing.T) {
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/chachapoly_test.go:155]: Same naming style here.

File: security/s2a/internal/crypter/chachapoly_test.go [hunk 20]
```diff
@@ -0,0 +1,214 @@
...
+			t.Fatalf("sender UpdateKey failed with: %v", err)
+		}
+		if err := receiver.updateKey(newKey); err != nil {
+			t.Fatalf("receiver UpdateKey failed with: %v", err)
+		}
+		testPOLYEncryptRoundtrip(sender, receiver, t)
+	}
+}
+
+func TestWycheProofTestVectorsCCP(t *testing.T) {
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/chachapoly_test.go:172]: s/CCP/ChachaPoly

File: security/s2a/internal/crypter/chachapoly.go (hunk 9 above)
Comment (davisgu) [reply security/s2a/internal/crypter/chachapoly.go:93]: common.go -- also seems to be used in crypter/chacha20poly1305.go library. The alternative is to rename the predefined gcmTagSize to tagSize and use it throughout the files?

File: security/s2a/internal/crypter/chachapoly_test.go [hunk 21]
```diff
@@ -0,0 +1,214 @@
...
+		t.Run(fmt.Sprintf("%d/%s", test.ID, test.Desc), func(t *testing.T) {
+			// Test encryption and decryption for CHACHA-POLY.
+			sender, receiver := getPOLYCryptoPair(test.Key, t)
+			testPOLYEncryptionDecryption(sender, receiver, &test, t)
+		})
+	}
+}
+
+// Test CHACHA-POLY with RFC test vectors.
+func TestCHACHAPOLYRFC(t *testing.T) {
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/chachapoly_test.go:187]: Same naming style.

**Comment by matthewstevenson88** [comment]
Thanks for the PR Davis! I've done a first pass of comments. Please send the changes to AES-GCM as a separate PR.

File: security/s2a/internal/crypter/aesgcm_test.go (hunk 1 above)
Comment (Ryanfsdf) [reply security/s2a/internal/crypter/aesgcm_test.go:26]: Strange, it could be that we're using different go linter versions.

File: security/s2a/internal/crypter/aesgcm_test.go [hunk 22]
```diff
@@ -40,11 +41,11 @@ func getGCMCryptoPair(key []byte, t *testing.T) (s2aAeadCrypter, s2aAeadCrypter)
 
 func isFailure(result string, err error, got, expected []byte) bool {
```
Comment (Ryanfsdf) [inline security/s2a/internal/crypter/aesgcm_test.go:42]: If this is going to be used in the ChaChaPoly test, we should move this over the the `wycheproofutil.go` or `testutil/common.go` file

Comment (davisgu) [reply security/s2a/internal/crypter/aesgcm_test.go:42]: That was my thought as well, but when I moved it to common.go, I recieved an interesting error:

`go: finding module for package google.go.lang/grpc/grpc-go/security/s2a/internal/crypter/testutil
common.go:6:2: cannot find module providing package google.go.lang/grpc/grpc-go/security/s2a/internal/crypter/testutil: unrecognized import path "google.go.lang/grpc/grpc-go/security/s2a/internal/crypter/testutil": https fetch: Get "https://google.go.lang/grpc/grpc-go/security/s2a/internal/crypter/testutil?go-get=1": dial tcp: lookup google.go.lang on 127.0.0.1:53: no such host`

It seems that it couldn't locate 'google.go.lang/grpc/grpc-go/security/s2a/internal/crypter/testutil'?

Comment (Ryanfsdf) [reply security/s2a/internal/crypter/aesgcm_test.go:42]: `testutil` is already imported in your `chachapoly_test.go` file, so I'm not sure if it's an import issue. When you moved it to `testutil/common.go`, did you export the function?
ie.
`isFailure`->`IsFailure`

I just tried moving the function over to `common.go` and it's working for me.

Comment (davisgu) [reply security/s2a/internal/crypter/aesgcm_test.go]: I moved it to wycheproofutil and worked fine.

File: security/s2a/internal/crypter/chachapoly.go (hunk 9 above)
Comment (matthewstevenson88) [reply security/s2a/internal/crypter/chachapoly.go:92]: Please rename all of them `tagSize`.

File: security/s2a/internal/crypter/chachapoly.go (hunk 8 above)
Comment (matthewstevenson88) [reply security/s2a/internal/crypter/chachapoly.go]: Yes please!

File: security/s2a/internal/crypter/aesgcm_test.go (hunk 1 above)
Comment (matthewstevenson88) [reply security/s2a/internal/crypter/aesgcm_test.go]: As long as it was just the linter, that's fine.

File: security/s2a/internal/crypter/chachapoly.go [hunk 23]
```diff
@@ -0,0 +1,106 @@
...
+package crypter
+
+import (
+	"crypto/cipher"
+	"fmt"
+
+	"golang.org/x/crypto/chacha20poly1305"
+)
+
+// Supported key sizes in bytes.
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/chachapoly.go:28]: s/sizes/size

File: security/s2a/internal/crypter/chachapoly.go [hunk 24]
```diff
@@ -0,0 +1,106 @@
...
+	// is necessary to restrict key updates to the same key length as the
+	// initial key.
+	keySize int
+}
+
+// newChachaPoly creates a Chacha-Poly crypter instance. Note that the key must be
+// 32 bytes.
+func newChachaPoly(key []byte) (s2aAeadCrypter, error) {
+	if len(key) != chacha20Poly1305KeySize {
+		return nil, fmt.Errorf("32 bytes, given: %d", len(key))
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/chachapoly.go:46]: Can you change `32` to `%d` and pass in `chacha20Poly1305KeySize`?

File: security/s2a/internal/crypter/chachapoly.go [hunk 25]
```diff
@@ -0,0 +1,106 @@
...
+type chachapoly struct {
+	aead cipher.AEAD
+	// keySize stores the size of the key which was initially used. This
+	// is necessary to restrict key updates to the same key length as the
+	// initial key.
+	keySize int
+}
+
+// newChachaPoly creates a Chacha-Poly crypter instance. Note that the key must be
+// 32 bytes.
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/chachapoly.go:43]: s/32/chacha20Poly1305KeySize

File: security/s2a/internal/crypter/chachapoly.go [hunk 26]
```diff
@@ -0,0 +1,106 @@
...
+// the ciphertext that will not be encrypted but will be authenticated. If dst
+// has enough capacity to hold these bytes, the ciphertext and the tag, no
+// allocation and copy operations will be performed. dst and plaintext may
+// fully overlap or not at all.
+func (s *chachapoly) encrypt(dst, plaintext, nonce, aad []byte) ([]byte, error) {
+	if len(nonce) != nonceSize {
+		return nil, fmt.Errorf("nonce size must be %d bytes. received: %d", nonceSize, len(nonce))
+	}
+	// If we need to allocate an output buffer, we want to include space for
+	// the tag to avoid forcing TLS record to reallocate as well.
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/chachapoly.go:66]: There's no mention of TLS records in this API, so please remove the mention here. Refer to the caller or similar, instead.

File: security/s2a/internal/crypter/chachapoly_test.go [hunk 27]
```diff
@@ -0,0 +1,214 @@
...
+package crypter
+
+import (
+	"fmt"
+	"testing"
+
+	"google.golang.org/grpc/security/s2a/internal/crypter/testutil"
+)
+
+// getChachaPolyCrypterPair outputs a sender/receiver pair on CHACHA-POLY.
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/chachapoly_test.go:28]: s/on CHACHA-POLY/of CHACHA-POLY AEAD crypters.

File: security/s2a/internal/crypter/chachapoly_test.go [hunk 28]
```diff
@@ -0,0 +1,214 @@
...
+			test.Key, ciphertext, test.Nonce, test.Aad, got, err, test.Plaintext)
+	}
+}
+
+func testChachaPolyEncryptRoundtrip(sender s2aAeadCrypter, receiver s2aAeadCrypter, t *testing.T) {
+	// Construct a dummy nonce.
+	nonce := make([]byte, nonceSize)
+
+	// Encrypt.
+	const plaintext = "This is plaintext."
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/chachapoly_test.go:75]: Is there any test where `plaintext` is empty? E.g. does one of the Wycheproof test vectors cover this case?

File: security/s2a/internal/crypter/chachapoly_test.go [hunk 29]
```diff
@@ -0,0 +1,214 @@
...
+	decryptedPlaintext, err := receiver.decrypt(ciphertext[:0], ciphertext, nonce, nil)
+	if err != nil {
+		t.Fatalf("Decrypt(%v, %v, %v, nil) failed, err = %v", ciphertext[:0], ciphertext, nonce, err)
+	}
+	if string(decryptedPlaintext) != plaintext {
+		t.Fatalf("Decrypt(%v, %v, %v, nil) = %v, want %v", ciphertext[:0], ciphertext, nonce, decryptedPlaintext, plaintext)
+	}
+}
+
+// Test encrypt and decrypt using an invalid key size.
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/chachapoly_test.go:93]: Doesn't this test the constructor, and not encrypt/decrypt?

File: security/s2a/internal/crypter/common.go [hunk 30]
```diff
@@ -1,9 +1,9 @@
 package crypter
 
 const (
-	// gcmTagSize is the tag size in bytes. From crypto/cipher/gcm.go in the Go
-	// crypto library.
-	gcmTagSize = 16
+	// tagSize is the tag size in bytes. From crypto/cipher/gcm.go and
+	//crypto/chacha20poly1305  in the Go crypto library
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/common.go:5]: Please remove the library references and simply say `the tag size in bytes for AES-128-GCM-SHA256, AES-256-GCM-SHA384, and CHACHA20-POLY1305-SHA256.`.

File: security/s2a/internal/crypter/testutil/wycheproofutil.go [hunk 31]
```diff
@@ -47,3 +48,8 @@ func ParseWycheProofTestVectors(jsonFilePath string, shouldFilter func(TestGroup
 
 	return testVectors
 }
+
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/testutil/wycheproofutil.go:51]: Please add a short comment that describes what this function does.

File: security/s2a/internal/crypter/aesgcm_test.go (hunk 22 above)
Comment (matthewstevenson88) [reply security/s2a/internal/crypter/aesgcm_test.go]: @Ryanfsdf @davisgu Can this be resolved? Or is there anything left unaddressed?

**Comment by matthewstevenson88** [comment]
Thanks Davis! I've left a few more comments.

File: security/s2a/internal/crypter/aesgcm_test.go (hunk 22 above)
Comment (Ryanfsdf) [reply security/s2a/internal/crypter/aesgcm_test.go]: I think it just needs to be removed from `aesgcm_test.go` since it was moved to `wycheproofutil.go`

File: security/s2a/internal/crypter/chachapoly_test.go (hunk 28 above)
Comment (davisgu) [reply security/s2a/internal/crypter/chachapoly_test.go:75]: Can confirm there are multiple test cases where the plaintext is empty, in combination with other edge cases.

File: security/s2a/internal/crypter/aesgcm_test.go (hunk 22 above)
Comment (davisgu) [reply security/s2a/internal/crypter/aesgcm_test.go]: ^will create a new pr for the aesgcm changes

File: security/s2a/internal/crypter/chachapoly_test.go (hunk 28 above)
Comment (matthewstevenson88) [reply security/s2a/internal/crypter/chachapoly_test.go:75]: SG. Thanks! :)

File: security/s2a/internal/crypter/chachapoly.go [hunk 32]
```diff
@@ -0,0 +1,106 @@
...
+type chachapoly struct {
+	aead cipher.AEAD
+	// keySize stores the size of the key which was initially used. This
+	// is necessary to restrict key updates to the same key length as the
+	// initial key.
+	keySize int
+}
+
+// newChachaPoly creates a Chacha-Poly crypter instance. Note that the key must be
+// chacha20Poly1305KeySize bytes.
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/chachapoly.go:43]: s/bytes/bytes in length

**Comment by matthewstevenson88** [comment]
Thanks for the PR Davis! I've left one last nit and I approved. I'm adding @cesarghali for Go readability.
